src/poetry/mixology/version_solver.py

```diff
@@ -23,12 +24,43 @@
     from poetry.core.packages.package import Package
     from poetry.core.packages.project_package import ProjectPackage
 
+    from poetry.packages import DependencyPackage
     from poetry.puzzle.provider import Provider
 
 
 _conflict = object()
 
 
+class DependencyCache:
+    """
+    A cache of the valid dependencies.
+
+    The key observation here is that during the search - except at backtracking
+    - once we have decided that a dependency is invalid, we never need check it
+    again.
+    """
+
+    def __init__(self, provider: Provider):
+        self.provider = provider
+        self.cache: dict[str, list[Package]] = {}
+
+    @functools.lru_cache(maxsize=128)
```

Comment:
Is there a real advantage/disadvantage of limiting the size here? ie. when do we start seeing issues?

Reply:
I don't think it's a big deal either way.  

In the kind of edge case that this MR targets, we can get a large number of not-quite-the-same dependencies that differ only because they have different constraints, and I don't think that it's useful to keep the earlier versions of these dependencies cached.  But I don't suppose it does much harm either.